include/core/lf_token.h

```diff
@@ -0,0 +1,362 @@
+    void* (*copy_constructor) (void* value);
+} token_type_t;
+
+/**
+ * Token type for dynamically allocated arrays and structs sent as messages.
+ *
+ * This struct is the wrapper around the dynamically allocated memory
+ * that carries the message.  The message can be an array of values,
+ * where the size of each value is element_size (in bytes). If it is
+ * not an array, or is not to be treated as an array, the length == 1.
+ *
+ * In the C LF target, a type for an output that ends in '*' or '[]' is
+ * treated specially. The value carried by the output is assumed
+ * to be in dynamically allocated memory, and, using reference
+ * counting, after the last downstream reader of the value has
+ * finished, the memory will be freed.  To prevent this freeing
+ * from occurring, the output type can be specified using the
+ * syntax {= type* =}; this will not be treated as dynamically
+ * allocated memory. Alternatively, the programmer can give a typedef
+ * in the preamble that masks the trailing *.
+ */
+typedef struct lf_token_t {
+    /** Pointer to dynamically allocated memory containing a message. */
+    void* value;
+    /** Length of the array or 1 for a non-array. */
+    size_t length;
+    /** Pointer to the port or action defining the type of the data carried. */
+    token_type_t* type;
+    /** The number of times this token is on the event queue. */
+    size_t ref_count;
+    /** Convenience for constructing a temporary list of tokens. */
+    struct lf_token_t* next;
+} lf_token_t;
+
+/**
+ * A record of the subset of channels of a multiport that have present inputs.
+ */
+typedef struct lf_sparse_io_record_t {
+	int size;  			// -1 if overflowed. 0 if empty.
+	size_t capacity;    // Max number of writes to be considered sparse.
+	size_t present_channels[];  // Array of channel indices that are present.
+} lf_sparse_io_record_t;
+
+/**
+ * @brief Base type for ports (lf_port_base_t) and actions (trigger_t), which can carry tokens.
+ * The structs lf_port_base_t and trigger_t should start with an instance of this struct
+ * so that they can be cast to this struct to access these fields in a uniform way.
+ */
+typedef struct token_template_t {
+	/** Instances of this struct can be cast to token_type_t. */
+	token_type_t type;
+	lf_token_t* token;
+    size_t length;       // The token's length, for convenient access in reactions.
+} token_template_t;
+
+/**
+ * @brief Base type for ports.
+ * Port structs are customized types because their payloads are type
+ * specific. This struct represents their common features. Given any
+ * pointer to a port struct, it can be cast to lf_port_base_t and then
+ * these common fields can be accessed.
+ * IMPORTANT: If this is changed, it must also be changed in
+ * CPortGenerator.java generateAuxiliaryStruct().
+ */
+typedef struct lf_port_base_t {
+	token_template_t tmplt;               // Type and token information (template is a C++ keyword).
```

Comment:
I have heard from @erlingrj that ports consume a substantial amount of memory in the C target. Will this addition to `lf_port_base_t` cause problems on resource-constrained platforms? Or could it cause more general performance problems on platforms with caches?

Reply:
I don't think `lf_port_base_t` has added any fields to the port structs.  It just rearranges them so that the common fields across ports always come first, so any port struct can safely be cast to `lf_port_base_t`.

Reply:
Okay, maybe I misunderstood.

I was assuming that because `port_base_t` now includes this `token_template_t tmplt` field which it did not have before, that would mean that all ports have the field, even if they do not use tokens. `token_template_t tmplt` alone looks like it should consume 5 quadwords, whereas the part of `port_base_t` that existed before this PR is 3 quadwords and the total memory usage that was reported in the [NGRES23 paper](https://github.com/icyphy/lf-pubs/blob/master/ngres23/ngres23.tex) was 5 quadwords I think.

Reply:
I observed that ports need memory. I think we were limited to 10 Reactors in the Big benchmark from Savina on the nrf52. This is because there was an all-to-all connection topology. So we had, in fact a total of a 100 ports in that program.

Reply:
@petervdonovan is correct here and the new design does use more memory for each port. I would prefer to get this merged and deal with this in a separate PR. I've opened issue #151 for this.